Auth0/WebViewProvider.swift

```diff
@@ -0,0 +1,163 @@
+        self.request = URLRequest(url: authorizeURL)
+        self.redirectURL = redirectURL
+        self.callback = callback
+        self.viewController = viewController
+        self.viewController.modalPresentationStyle = modalPresentationStyle
+        
+        super.init()
+        if !defaultSchemesSupportedByWKWebview.contains(redirectURL.scheme!) {
+            self.setupWebViewWithCustomScheme()
+        } else {
+            self.setupWebViewWithHTTPS()
+        }
+    }
+
+    private func setupWebViewWithCustomScheme() {
+        let configuration = WKWebViewConfiguration()
+        configuration.setURLSchemeHandler(self, forURLScheme: redirectURL.scheme!)
+        self.webview = WKWebView(frame: .zero, configuration: configuration)
+        self.viewController.view = webview
+        webview.navigationDelegate = self
+    }
+
+    private func setupWebViewWithHTTPS() {
+        self.webview = WKWebView(frame: .zero)
+        self.viewController.view = webview
+        webview.navigationDelegate = self
+    }
+    
+    func start() {
+        self.webview.load(self.request)
+        self.topViewController?.present(self.viewController, animated: true)
+    }
+
+    func finish(with result: WebAuthResult<Void>) {
+        DispatchQueue.main.async { [weak webview, weak viewController, callback] in
+            webview?.removeFromSuperview()
+            guard let presenting = viewController?.presentingViewController else {
+                let error = WebAuthError(code: .unknown("Cannot dismiss WKWebView"))
+                return callback(.failure(error))
+            }
+            presenting.dismiss(animated: true) {
+                callback(result)
+            }
+        }
+    }
+    
+    public override var description: String {
+        return String(describing: WKWebView.self)
+    }
+}
+
+/// Handling Custom Scheme Callbacks
+extension WebViewUserAgent: WKURLSchemeHandler {
+    func webView(_ webView: WKWebView, start urlSchemeTask: any WKURLSchemeTask) {
+        _ = TransactionStore.shared.resume(urlSchemeTask.request.url!)
+        let error = NSError(domain: WebViewUserAgent.customSchemeRedirectionSuccessMessage, code: 200, userInfo: [
+            NSLocalizedDescriptionKey: "WebViewProvider: WKURLSchemeHandler: Succesfully redirected back to the app"
+        ])
+        urlSchemeTask.didFailWithError(error)
+    }
+    
+    func webView(_ webView: WKWebView, stop urlSchemeTask: any WKURLSchemeTask) {
+        let error = NSError(domain: WebViewUserAgent.customSchemeRedirectionFailureMessage, code: 400, userInfo: [
+            NSLocalizedDescriptionKey: "WebViewProvider: WKURLSchemeHandler: Webview Resource Loading has been stopped"
+        ])
+        urlSchemeTask.didFailWithError(error)
+        self.finish(with: .failure(WebAuthError(code: .webViewResourceLoadingStopped)))
```

Comment:
Should this error case be `userCancelled`?

Reply:
yes, updated accordingly